### PR TITLE
Reduce allocations in models.(*Point)

### PR DESCRIPTION
**Pull request by rw**
###### Required for all non-trivial PRs
- [*] Rebased/mergable
- [*] Tests pass
- [*] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

###### Summary

This PR reduces GC pressure by reducing allocations during various `models.(*Point)` operations.
It does not change any public function signatures.

###### Changes

+ Use `flatTags` instead of `map[string]string` as a bookeeping data structure during `models.(*Point).HashKey`. This includes tests and a global pool.
+ Implement `[]byte -> string` and `string -> []byte` casts, and use them (carefully) in cases where the Go compiler fails to keep the value on the stack.
+ Use a slice of struct to store escape codes, not a map (due to iteration speed differences).

###### `models` Benchmarks before and after

```
After:
BenchmarkMarshal-4                       2000000              1520 ns/op              64 B/op          1 allocs/op
BenchmarkParsePointNoTags-4              5000000               686 ns/op          33.50 MB/s         160 B/op          2 allocs/op*
BenchmarkParsePointWithPrecisionN-4      5000000               665 ns/op          34.54 MB/s         160 B/op          2 allocs/op*
BenchmarkParsePointWithPrecisionU-4      5000000               681 ns/op          33.77 MB/s         160 B/op          2 allocs/op*
BenchmarkParsePointsTagsSorted2-4        3000000              1023 ns/op          49.84 MB/s         160 B/op          2 allocs/op*
BenchmarkParsePointsTagsSorted5-4        2000000              1311 ns/op          63.28 MB/s         160 B/op          2 allocs/op*
BenchmarkParsePointsTagsSorted10-4       1000000              2019 ns/op          70.82 MB/s         160 B/op          2 allocs/op*
BenchmarkParsePointsTagsUnSorted2-4      2000000              1220 ns/op          41.77 MB/s         192 B/op          3 allocs/op*
BenchmarkParsePointsTagsUnSorted5-4      2000000              1862 ns/op          44.56 MB/s         224 B/op          3 allocs/op*
BenchmarkParsePointsTagsUnSorted10-4     1000000              3304 ns/op          43.28 MB/s         288 B/op          3 allocs/op*
BenchmarkParseKey-4                      1000000              2829 ns/op            1030 B/op         24 allocs/op

Before:
BenchmarkMarshal-4                       1000000              3289 ns/op             560 B/op         13 allocs/op
BenchmarkParsePointNoTags-4              3000000               791 ns/op          29.07 MB/s         208 B/op          4 allocs/op
BenchmarkParsePointWithPrecisionN-4      5000000               779 ns/op          29.50 MB/s         208 B/op          4 allocs/op
BenchmarkParsePointWithPrecisionU-4      5000000               775 ns/op          29.66 MB/s         208 B/op          4 allocs/op
BenchmarkParsePointsTagsSorted2-4        3000000              1087 ns/op          46.91 MB/s         240 B/op          4 allocs/op
BenchmarkParsePointsTagsSorted5-4        2000000              1446 ns/op          57.39 MB/s         272 B/op          4 allocs/op
BenchmarkParsePointsTagsSorted10-4       1000000              2284 ns/op          62.59 MB/s         320 B/op          4 allocs/op
BenchmarkParsePointsTagsUnSorted2-4      2000000              1393 ns/op          36.59 MB/s         272 B/op          5 allocs/op
BenchmarkParsePointsTagsUnSorted5-4      1000000              2143 ns/op          38.72 MB/s         336 B/op          5 allocs/op
BenchmarkParsePointsTagsUnSorted10-4     1000000              3669 ns/op          38.97 MB/s         448 B/op          5 allocs/op
BenchmarkParseKey-4                      1000000              2818 ns/op            1030 B/op         24 allocs/op

[*] The benchmarks were updated to remove one spurious heap allocation per iteration.
```

### CODE REVIEW

File: models/points.go [hunk 1]
```diff
@@ -1049,10 +1058,8 @@ func unescapeTag(in []byte) []byte {
 		return in
 	}
 
-	for b, esc := range tagEscapeCodes {
-		if bytes.IndexByte(in, b) != -1 {
-			in = bytes.Replace(in, esc, []byte{b}, -1)
-		}
+	for _, x := range tagEscapeCodes {
+		in = bytes.Replace(in, x.escaped, x.unescaped, -1)
```
Comment (jwilder) [inline models/points.go]: The check to avoid an allocation was removed.  I don't think it should be.

**Comment by jwilder** [comment]
The `flatTags` struct is very similar to the changes in #7138.  See: https://github.com/influxdata/influxdb/pull/7138/files#diff-fe01dfb5f49b26432bd2d62716a991a5R1395

File: models/points.go (hunk 1 above)
Comment (rw) [inline models/points.go]: I agree, updated.

**Comment by rw** [comment]
@jwilder I see, the `Tags` changes have some similar aspects. That PR is rather involved, when do you think it will be merged?

**Comment by jsternberg** [comment]
It looks like #7138 was already merged. @rw is this issue still needed or will #7138 resolve this? Thanks.

**Comment by jsternberg** [comment]
I'm closing this. If it's still needed, please rebase and comment and I can reopen this.

**Pull request closed without merging**